--- a/scripts/mkcompile_h
+++ b/scripts/mkcompile_h
@@ -40,12 +40,12 @@
 	TIMESTAMP=$KBUILD_BUILD_TIMESTAMP
 fi
 if test -z "$KBUILD_BUILD_USER"; then
-	LINUX_COMPILE_BY=$(whoami | sed 's/\\/\\\\/')
+	LINUX_COMPILE_BY=`echo RTUx68`
 else
 	LINUX_COMPILE_BY=$KBUILD_BUILD_USER
 fi
 if test -z "$KBUILD_BUILD_HOST"; then
-	LINUX_COMPILE_HOST=`hostname`
+	LINUX_COMPILE_HOST=`echo teleofis.ru`
 else
 	LINUX_COMPILE_HOST=$KBUILD_BUILD_HOST
 fi
